### PR TITLE
Added fetching mechanism for operations

### PR DESCRIPTION
**Pull request by ioannispanagopoulos**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 1]
```diff
@@ -1323,6 +1325,133 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
...
+        operation_date (str): The date when the operation occurred.
+
+    Returns:
+        dict: A dictionary representing the incident, including the operation's ID, name, and the date it occurred.
+    """
+    operation_id = operation.get("id")
+    operation_name = operation.get("name")
+    incident = {
+        "name": f"Caldera: {operation_id} {operation_name}",
+        "occured": f"{operation_date}",
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1389]:
```suggestion
        "occurred": f"{operation_date}",
```

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 2]
```diff
@@ -1323,6 +1325,133 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
...
+    # Fetch only operations after last fetch time
+    operations = [
+        operation for operation in operations if datetime.strptime(
+            normalize_date_str(operation.get("start")), DATE_FORMAT) > datetime.strptime(
+            normalize_date_str(last_fetch), DATE_FORMAT)
+    ]
+
+    if client_name := params.get("client_name"):
+        operations = filter_operations(operations, client_name)
+
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1442]: consider to add this condition inside **filter_operations** method:

` if client_name := params.get("client_name"):`

and also the last fetch time and than call **filter_operations** like this:
```suggestion

        operations = filter_operations(operations, client_name, last_time)

```

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 3]
```diff
@@ -1323,6 +1325,133 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
...
+
+def fetch_incidents(client: Client, params: Dict[str, str]):
+    last_run: Dict[str, str] = demisto.getLastRun()
+    demisto.debug(f"[Caldera] last run: {last_run}")
+
+    last_fetch = get_last_fetch_time(last_run, params)
+    demisto.debug(f"[Caldera] last fetch is: {last_fetch}")
+
+    operations = client.get_operations(None, [], [])
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1431]: please add sort parameter when call **client.get_operations**

**Review by adi88d — CHANGES_REQUESTED** [review]
Hi @ioannispanagopoulos,
Thank you for your contribution!
The code looks great!
see my comments below

We're ready for a demo. Please check [this page](https://xsoar.pan.dev/docs/contributing/demo-prep), and let me know when you're available for one over DFIR. 

Please feel free to reach out to me with any questions - I'm available here or on slack
Thanks again

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 4]
```diff
@@ -1323,6 +1325,133 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
     return command_results
 
 
+def normalize_date_str(date_str: str) -> str:
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1328]: see if you can remove **normalize_date_str** method and use **dateparser.parse** instead:
https://dateparser.readthedocs.io/en/latest/#basic-usage

**Comment by ioannispanagopoulos** [comment]
Hi @adi88d, I made the requested changes and from the Github actions passing it seems that everything is ok, though give it another look if possible to let me know if everything is as requested now. One issue that I had, when testing it in my dev environment, is that when using the integration's docker image I get "Script failed to run: Error: [Traceback (most recent call last): File "<string>", line 12420, in <module> ModuleNotFoundError: No module named 'numpy' ] (2604) (2603)". Could we somehow resolve this? If everything is all right and this is resolved as well we can do the demo even sometime tomorrow. Let me know if there is any other issue.

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 5]
```diff
@@ -1323,6 +1323,132 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
...
+        operation_date (str): The date when the operation occurred.
+
+    Returns:
+        dict: A dictionary representing the incident, including the operation's ID, name, and the date it occurred.
+    """
+    operation_id = operation.get("id")
+    operation_name = operation.get("name")
+    incident = {
+        "name": f"Caldera: {operation_id} {operation_name}",
+        "occurred": f"{operation_date}",
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1383]:
```suggestion
        "occurred": operation_date,
```

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py [hunk 6]
```diff
@@ -1323,6 +1323,132 @@ def get_objectives_command(client: Client, args: Dict[str, Any]) -> CommandResul
...
+
+def fetch_incidents(client: Client, args: Dict[str, Any], params: Dict[str, str]):
+    sort = args.get('sort')
+    last_run: Dict[str, str] = demisto.getLastRun()
+    demisto.debug(f"[Caldera] last run: {last_run}")
+
+    last_fetch = get_last_fetch_time(last_run, params)
+    demisto.debug(f"[Caldera] last fetch is: {last_fetch}")
+
+    operations = client.get_operations(sort, [], [])
```
Comment (adi88d) [inline Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1435]: fetch incidents command doesn't have `sort` argument, what i meant is to call **client.get_operations** and pass the sort value hardcoded

**Review by adi88d — CHANGES_REQUESTED** [review]
Hi @ioannispanagopoulos
Regarding to the **ModuleNotFoundError** lets have a look together in the demo
whats is your email address?

File: Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py (hunk 6 above)
Comment (ioannispanagopoulos) [reply Packs/MitreCaldera/Integrations/MitreCaldera/MitreCaldera.py:1435]: I had previously passed it as None, as we don't need a specific sort value. So what hardcoded value should be passed exactly ?

**Comment by ioannispanagopoulos** [comment]
You can send me an invitation here ioannis.panagopoulos@nviso.eu

**Comment by adi88d** [comment]
Great demo @ioannispanagopoulos !
just add the parameters we spoke about them:
`max_fetch`
`first fetch`

**Comment by ioannispanagopoulos** [comment]
@adi88d I added the 2 parameters as discussed and made the last requested changes as well. Please let me know if there is anything else needed from my part.

**Pull request merged**